File: deepseek/numerical_integration.py
import math
import numpy as np

def exact_integral(a: float, b: float) -> float:
    """Compute the exact integral of sin(x) from a to b"""
    return -math.cos(b) + math.cos(a)

def riemann_sum(a: float, b: float, n: int = 1_000_000) -> float:
    """
    Compute Riemann sum approximation of sin(x) from a to b
    using the midpoint rule for better accuracy
    
    Args:
        a: lower bound
        b: upper bound
        n: number of rectangles (default: 1,000,000)
    
    Returns:
        Approximate integral value
    """
    dx = (b - a) / n  # Width of each rectangle
    x = np.linspace(a + dx/2, b - dx/2, n)  # Midpoints of all intervals
    return np.sum(np.sin(x) * dx)

def main():
    a = -math.pi
    b = 2/3 * math.pi
    n = 1_000_000  # Number of rectangles
    
    # Compute Riemann sum approximation
    approx_area = riemann_sum(a, b, n)
    
    # Compute exact integral
    exact_area = exact_integral(a, b)
    
    # Calculate errors
    abs_error = abs(approx_area - exact_area)
    rel_error = abs_error / abs(exact_area) * 100
    
    # Print results
    print(f"Calculating area of sin(x) from {a:.6f} to {b:.6f}\n")
    print(f"Riemann sum approximation ({n:,} rectangles): {approx_area:.12f}")
    print(f"Exact integral value: {exact_area:.12f}")
    print(f"Absolute error: {abs_error:.12f}")
    print(f"Relative error: {rel_error:.6f} %")

if __name__ == "__main__":
    main()

File: deepseek/numerical_integration.m
function sinx_riemann_area()
    % Main function to compute and validate the area under sin(x)
    
    % Integration bounds
    a = -pi;
    b = (2/3)*pi;
    
    % Number of rectangles (increase for better accuracy)
    n = 1000000;
    
    % Compute Riemann sum approximation
    approx_area = riemann_sum(a, b, n);
    
    % Compute exact integral value
    exact_area = exact_integral(a, b);
    
    % Calculate errors
    abs_error = abs(approx_area - exact_area);
    rel_error = (abs_error / abs(exact_area)) * 100;
    
    % Display results
    fprintf('Calculating area of sin(x) from %.6f to %.6f\n\n', a, b);
    fprintf('Riemann sum approximation (%,d rectangles): %.12f\n', n, approx_area);
    fprintf('Exact integral value: %.12f\n', exact_area);
    fprintf('Absolute error: %.12f\n', abs_error);
    fprintf('Relative error: %.6f %%\n\n', rel_error);
    
    % Optional visualization
    visualize(a, b);
end

function area = riemann_sum(a, b, n)
    % Compute Riemann sum using midpoint rule
    dx = (b - a)/n;          % Width of each rectangle
    x = a + dx/2:dx:b-dx/2;  % Midpoints of all intervals
    area = sum(sin(x)*dx);   % Sum of areas of all rectangles
end

function integral = exact_integral(a, b)
    % Compute exact integral of sin(x) from a to b
    integral = -cos(b) + cos(a);
end

function visualize(a, b)
    % Visualize the function and shaded area
    figure;
    x = linspace(a, b, 1000);
    y = sin(x);
    
    plot(x, y, 'b', 'LineWidth', 2);
    hold on;
    area(x, y, 'FaceColor', [0.7 0.8 1], 'EdgeColor', 'none');
    plot(x, zeros(size(x)), 'k--'); % x-axis
    title(sprintf('Area under sin(x) from %.2f to %.2f', a, b));
    xlabel('x');
    ylabel('sin(x)');
    grid on;
    hold off;
end
